Copied over fixed parameters to new models

brechmos-stsci opened the pull request:
The fixed parameter attribute of the model was not copied over when stripping and adding units. This is now fixed and a test was added based on the code in the issue below.

Fixes https://github.com/astropy/specutils/issues/357

nmearl commented:
@brechmos-stsci Thanks for looking into this. I think we should also be sure that things like the `tied` and `bounds` parameters are all preserved.

brechmos-stsci commented:
Are there any others besides fixed, tied and bounds?  I'll add to this PR.

nmearl commented:
I believe that those three are the only things that influence the parameter behavior in model fitting.

brechmos-stsci commented:
Yeesh, this is going to get more complex.  It looks like I am going to have to do the same thing to the bounds (remove units on bounds, do fit, add units back on bounds).

eteq commented:
Quick thought about this: in the future (possibly even the upcoming astropy) there'll be options for more model attributes like `priors`, possibly some metadata, etc.  In principal that also needs to be transferred over.

I haven't looked close enough yet to know the details, but ideally this copying machinery would copy over *everything*.  Is that possible using e.g. `deepcopy` or similar?

brechmos-stsci commented:
I need to remember, but I think the problem is that once we do a `deepcopy` it is not possible to set some parameters as they are attributes and don't have setters. So, I don't **think** it is that easy.

brechmos-stsci commented:
Two options, either `_constraints` or loop through `constraints` and use `setattr` (per @eteq)

brechmos-stsci commented:
Ok, for this PR, I moved the convert-and-strip-units to a separate function as it was needed for the bounds constraint part. This is now in a function `_convert_and_dequantify`. As well, after a discussion with @eteq, I am copying the whole `model._constraints` dict so that things like `tied`, `bounds` etc and any future things in there will be copied to the new models.  I added some tests to confirm that these are being copied over correctly.

So.... this should be ready for review and merge. This is kinda necessary as technically the fitting isn't working properly (without `fixed` etc). 

@nmearl @eteq

eteq commented on specutils/fitting/fitmodels.py:
```diff
@@ -212,6 +212,44 @@ def _fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
+
+    """
+
+    if poss_quantity is None or isinstance(poss_quantity, (float, int)):
```
You might want to add an `isinstance(poss_quantity, Quantity)` clause here. It's *possible* that in the future someone (or a future version of `astropy,modeling` would see Quantity's coming in directly...

brechmos-stsci replied:
Uh, I don't quite understand. For that line we are specifically checking for *not* Quantities. is that what you meant?

eteq commented on specutils/fitting/fitmodels.py:
```diff
@@ -212,6 +212,44 @@ def _fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
+    if poss_quantity is None or isinstance(poss_quantity, (float, int)):
+        return poss_quantity
+
+    if hasattr(poss_quantity, 'quantity') and poss_quantity.quantity is not None:
```
I think my comment above might have better been here.  I think I meant a leading 
```
if isinstance(poss_quantity, Quantity):
    q = poss_quantity
elif ...
```

brechmos-stsci replied:
I made a small change to include that check.
